src/main/java/xbot/common/command/BaseRobot.java

```diff
@@ -252,62 +237,63 @@ public void teleopPeriodic() {
      */
     public void testPeriodic() {
     }
+
+    double outsidePeriodicStart = 0;
     
     protected void sharedPeriodic() {
-        outsidePeriodicMonitor.stop();
-        schedulerMonitor.start();
-        xScheduler.run();
-        schedulerMonitor.stop();
-
-        batteryVoltage.set(RobotController.getBatteryVoltage());
-        
-        brownoutLatch.setValue(RobotController.isBrownedOut());
-        
-        loopCycleCounter++;
-        double timeSinceLastLog = XTimer.getFPGATimestamp() - lastFreqCounterResetTime;
-        if(lastFreqCounterResetTime <= 0) {
-            lastFreqCounterResetTime = XTimer.getFPGATimestamp();
-        }
-        else if(timeSinceLastLog >= frequencyReportInterval.get()) {
-            double loopsPerSecond = loopCycleCounter / timeSinceLastLog; 
-            
-            loopCycleCounter = 0;
-            lastFreqCounterResetTime = XTimer.getFPGATimestamp();
-            
-            log.info("Robot loops per second: " + loopsPerSecond);
+        double outsidePeriodicEnd = getPerformanceTimestampInMs();
+        Logger.getInstance().recordOutput("OutsidePeriodicMs", outsidePeriodicEnd - outsidePeriodicStart);
+        // Get a fresh data frame from all top-level components (typically large subsystems or shared sensors)
+        double dataFrameStart = getPerformanceTimestampInMs();
+        for (DataFrameRefreshable refreshable : dataFrameRefreshables) {
+            refreshable.refreshDataFrame();
         }
+        double dataFrameEnd = getPerformanceTimestampInMs();
+        Logger.getInstance().recordOutput("RefreshDataFrameMs", dataFrameEnd - dataFrameStart);
+
+        double schedulerStart = getPerformanceTimestampInMs();
+        xScheduler.run();
+        double schedulerEnd = getPerformanceTimestampInMs();
+        Logger.getInstance().recordOutput("SchedulerMs", schedulerEnd - schedulerStart);
         
-        outsidePeriodicMonitor.start();
+        outsidePeriodicStart = getPerformanceTimestampInMs();
     }
 
     
     @Override
     public void simulationInit() {
-        /*
-        webots = injectorComponent.webotsClient();
-        webots.initialize();
-
-         */
+        // TODO: Add something to detect replay vs Webots, and skip all of this if we're in replay mode.
```

Comment:
not for now, but instead of `if webots` maybe should have an enum of simutation modes?